src/main/java/org/breedinginsight/services/parsers/trait/TraitFileParser.java

```diff
@@ -0,0 +1,230 @@
+
+        ArrayList<Trait> traits = new ArrayList<>();
+        InputStreamReader in = new InputStreamReader(inputStream);
+
+        Iterable<CSVRecord> records = null;
+        try {
+            // withHeader for enum uses name() internally so we have to give string array instead
+            records = CSVFormat.DEFAULT
+                    .parse(in);
+        } catch (IOException e) {
+            log.error(e.getMessage());
+            throw new ParsingException("Error reading file");
+        }
+
+        Sheet excelSheet = convertCsvToExcel(records);
+        ExcelParser parser = new ExcelParser();
+        List<ExcelRecord> excelRecords = parser.parse(excelSheet, TraitFileColumns.getColumns());
+        return excelRecordsToTraits(excelRecords);
+
+    }
+
+    private List<Trait> excelRecordsToTraits(List<ExcelRecord> records) {
+        List<Trait> traits = new ArrayList<>();
+
+        for (ExcelRecord record : records) {
+
+            ProgramObservationLevel level = ProgramObservationLevel.builder()
+                    .name(parseExcelValueAsString(record, TraitFileColumns.TRAIT_LEVEL))
+                    .build();
+
+            // TODO: throw if not active/archived/empty
+            Boolean active = !parseExcelValueAsString(record, TraitFileColumns.TRAIT_STATUS).equals(TRAIT_STATUS_ARCHIVED);
+
+            Method method = Method.builder()
+                    .methodName(parseExcelValueAsString(record, TraitFileColumns.METHOD_NAME))
+                    .description(parseExcelValueAsString(record, TraitFileColumns.METHOD_DESCRIPTION))
+                    .methodClass(parseExcelValueAsString(record, TraitFileColumns.METHOD_CLASS))
+                    .formula(parseExcelValueAsString(record, TraitFileColumns.METHOD_FORMULA))
+                    .build();
+
+            // TODO: throw if not valid datatype
+            DataType dataType = null;
+
+            try {
+                dataType = DataType.valueOf(parseExcelValueAsString(record, TraitFileColumns.SCALE_CLASS).toUpperCase());
+            } catch (IllegalArgumentException e) {
+                log.error(e.getMessage());
+            }
+
+            // TODO: throw if bad format
+            List<BrApiScaleCategories> categories = parseListValue(parseExcelValueAsString(record, TraitFileColumns.SCALE_CATEGORIES)).stream()
+                    .map(value -> parseCategory(value))
+                    .collect(Collectors.toList());
+
+            Integer decimalPlaces = null;
+            Integer validValueMin = null;
+            Integer validValueMax = null;
+
+            try {
+                decimalPlaces = Integer.valueOf(parseExcelValueAsString(record, TraitFileColumns.SCALE_DECIMAL_PLACES));
+                validValueMin = Integer.valueOf(parseExcelValueAsString(record, TraitFileColumns.SCALE_LOWER_LIMIT));
+                validValueMax = Integer.valueOf(parseExcelValueAsString(record, TraitFileColumns.SCALE_UPPER_LIMIT));
```

Comment:
Ya, if blank shouldn't get a value. I think I should break out each of these into their own try/catch so nothing gets skipped over and throw if it's not a number so it's clear that there was an error in the expected type and nothing is silently accepted with a missing value.

Reply:
could you remove the try/catch that's here now, and let the NumberFormatException propagate to the caller?

Reply:
One thing I would lose by passing it up and not wrapping in the excelRecordsToTraits is saying which field it occurred for rather than the value that failed.

Reply:
@nickpalladino and I talked about this on slack.  Agreed that this private method should have individual try/catch blocks for each integer parsing, and if there's an exception, throw a new `ParsingException`

Reply:
Updated, added tests, and pushed changes.